clang/lib/3C/3C.cpp

```diff
@@ -313,39 +441,55 @@ _3CInterface::_3CInterface(const struct _3COptions &CCopt,
   GlobalProgramInfo.getPerfStats().startTotalTime();
 }
 
+bool _3CInterface::finishPass(bool HadNonDiagnosticError,
+                              bool IsLastPass) {
+  assert(!FinishedLastPass);
+
+  bool HadError = HadNonDiagnosticError;
+  for (auto &TU : ASTs)
+    HadError |= (TU->getDiagnostics().getClient()->getNumErrors() > 0);
+
+  if (IsLastPass || HadError) {
+    // We won't run any more passes, so we need to finish any diagnostic
+    // verification now.
+    FinishedLastPass = true;
+    LastPassFailed = HadError;
+    LastPassFailedAfterDiagnosticVerification = HadNonDiagnosticError;
```

Comment:
Flagging this section as confusing (not necessarily incorrect), because `LastPassFailed` includes diagnostic failures while `LastPassFailedAfterDiagnosticVerification` includes non-diagnostic failures. And also two different types of failure. A comment about the different types would help, here and/or in `finish3CAnalysis`, since that location has three consumers vars. Better names could also help, though the logic would have to work around the names.

I don't know if the confusion is enough to warrant a rewrite. Comment is probably enough.

Reply:
I think my in-progress cleanup will probably already help somewhat with this, but I'll also look into adding more comments based on what you said here.